src/main/java/com/google/devtools/build/remote/client/LogParserUtils.java

```diff
@@ -0,0 +1,155 @@
+
+/**
+ * Methods for printing log files.
+ */
+public class LogParserUtils {
+
+  private static final String DELIMETER =
+      "---------------------------------------------------------\n";
+
+  public static class ParamException extends Exception {
+
+    public ParamException(String message) {
+      super(message);
+    }
+  }
+
+  private String filename;
+
+  public LogParserUtils(String filename) {
+    this.filename = filename;
+  }
+
+  private String getFilename() throws ParamException {
+    if (filename.isEmpty()) {
+      throw new ParamException("This operation cannot be performed without specifying --grpc_log.");
+    }
+    return filename;
+  }
+
+
+  /**
+   * Attempt to find and print ExecuteResponse from the details of a log entry for a Watch call. If
+   * no Operation could be found in the Watch call responses, or an Operation was found but failed,
+   * a failure message is printed.
+   */
+  private static void printExecuteResponse(WatchDetails watch, PrintWriter out) throws IOException {
+    for (ChangeBatch cb : watch.getResponsesList()) {
+      for (Change ch : cb.getChangesList()) {
+        if (ch.getState() != State.EXISTS) {
+          continue;
+        }
+        Operation o = ch.getData().unpack(Operation.class);
+        if (o.getResultCase() == ResultCase.ERROR && o.getError().getCode() != Code.OK.value()) {
+          out.printf("Operation contained error: %s\n", o.getError().toString());
+          return;
+        } else if (o.getResultCase() == ResultCase.RESPONSE && o.getDone()) {
+          out.println("ExecuteResponse extracted:");
+          out.println(o.getResponse().unpack(ExecuteResponse.class).toString());
+          return;
+        }
+      }
+    }
+    out.println("Could not find ExecuteResponse in Watch call details.");
+  }
+
+  /**
+   * Print an individual log entry.
+   */
+  static void printLogEntry(LogEntry entry, PrintWriter out) throws IOException {
+    out.println(entry.toString());
+    if (entry.getDetails().getDetailsCase() == DetailsCase.WATCH) {
+      out.println("\nAttempted to extract ExecuteResponse from Watch call responses:");
+      printExecuteResponse(entry.getDetails().getWatch(), out);
+    }
+  }
+
+  /**
+   * Prints each entry out individually (ungrouped) and a message at the end for how many entries
+   * were printed/skipped.
+   */
+  private void printEntriesInOrder(OutputStream outStream) throws IOException, ParamException {
+    try (InputStream in = new FileInputStream(getFilename())) {
+      PrintWriter out =
+          new PrintWriter(new BufferedWriter(new OutputStreamWriter(outStream, UTF_8)), true);
+      LogEntry entry;
+      while ((entry = LogEntry.parseDelimitedFrom(in)) != null) {
+        printLogEntry(entry, out);
+        System.out.print(DELIMETER);
+      }
+    }
+  }
+
+  private void printEntriesGroupedByAction(OutputStream outStream)
+      throws IOException, ParamException {
+    ActionGrouping byAction = new ActionGrouping();
```

Comment:
Now, while this is fine, the ActionGrouping class has so little code it really seems unnecessary. Like, you just want to gather and rearrange some stuff one time. I don't think you need a class for it, it would fit in a small while loop -- gathering the entries into a MultiMap and sorting the values before printing. I think it would be simpler and more readable. Like, my general rule is that I need an object if I want to keep state around to be able to do something multiple times on that object. This is not the case here.

Reply:
I don't want to parse the grpc log if it's not needed. The logic on whether it's needed is specific to the action. currently actionGrouping is just scoped by the method, but I'm planning to make it a class member that's lazilly initialized.

Also, actionGrouping will grow to include more elaborate logic for tallying and understanding the logEvents.

(couldn't add tests comment was regarding the old way where printing to stdout was interleaved with the logic. Obviously I can test a method regardless of which class it's in)

I'm not sure I understand what you're trying to achieve, could you expand a bit?

Reply:
Ok, sure -- let me just see where this is going, then, and if after then next few changes I still think it could be simplified, we'll revisit this.

Reply:
Thank you!
Sorry again for giving this to you in this mid-state. I realized that I need V2 for the changes I'm making, but my V2 implementation depends on this one.
I will definitely redraw this if the end state still seems problematic.